app/src/dex/router/path_cache.rs

```diff
@@ -0,0 +1,90 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use parking_lot::Mutex;
+use penumbra_crypto::asset;
+use penumbra_storage::{StateDelta, StateRead};
+
+use super::Path;
+
+/// An entry in the path cache, representing a best known sub-path.
+pub(super) struct PathEntry<S: StateRead + 'static> {
+    /// The best known path to the intermediate asset.
+    pub path: Path<S>,
+    /// Whether the path is active, used to implement the SPFA optimization.
+    /// Newly extended paths are active.  By deactivating paths if all of their
+    /// possible extensions are suboptimal, we can avoid re-extending them in
+    /// each iteration.
+    pub active: bool,
+    /// The second best known path to the intermediate asset, used to record spill prices.
+    pub spill: Option<Path<S>>,
+}
+
+impl<S: StateRead + 'static> PathEntry<S> {
+    /// Update this entry with the new path, if it's better than the existing one.
+    pub fn update(&mut self, new_path: Path<S>) {
+        if new_path.price < self.path.price {
+            tracing::debug!(new_price = ?new_path.price, old_price = ?self.path.price, "updating path");
```

Comment:
Here and below, I didn't include info about the path itself in the tracing event, on the assumption that we'll be calling this in the context of a path-related span.